core/base/geometry/Geometry.cpp

```diff
@@ -481,59 +517,353 @@ T Geometry::magnitude(const T *o, const T *d) {
 }
 
 template <typename T>
-int Geometry::subtractVectors(const T *a, const T *b, T *out) {
-  out[0] = b[0] - a[0];
-  out[1] = b[1] - a[1];
-  out[2] = b[2] - a[2];
+int Geometry::subtractVectors(const T *a,
+                              const T *b,
+                              T *out,
+                              const int &dimension) {
+  for(int i = 0; i < dimension; ++i)
+    out[i] = b[i] - a[i];
+  return 0;
+}
+
+template <typename T>
+int Geometry::subtractVectors(const std::vector<T> &a,
+                              const std::vector<T> &b,
+                              std::vector<T> &out) {
+  out.resize(a.size());
+  return subtractVectors(a.data(), b.data(), out.data(), a.size());
+}
+
+template <typename T>
+int Geometry::addVectors(const T *a, const T *b, T *out, const int &dimension) {
+  for(int i = 0; i < dimension; ++i)
+    out[i] = b[i] + a[i];
+  return 0;
+}
+
+template <typename T>
+int Geometry::addVectors(const std::vector<T> &a,
+                         const std::vector<T> &b,
+                         std::vector<T> &out) {
+  out.resize(a.size());
+  return addVectors(a.data(), b.data(), out.data(), a.size());
+}
+
+template <typename T>
+int Geometry::multiAddVectors(const std::vector<std::vector<T>> &a,
+                              const std::vector<std::vector<T>> &b,
+                              std::vector<std::vector<T>> &out) {
+  out.resize(a.size());
+  for(unsigned int i = 0; i < a.size(); ++i)
+    addVectors(a[i], b[i], out[i]);
+  return 0;
+}
+
+template <typename T>
+int Geometry::multiAddVectorsFlatten(
+  const std::vector<std::vector<std::vector<T>>> &a,
+  const std::vector<std::vector<std::vector<T>>> &b,
+  std::vector<std::vector<T>> &out) {
+  std::vector<std::vector<T>> a_flatten, b_flatten;
+  multiFlattenMultiDimensionalVector(a, a_flatten);
+  multiFlattenMultiDimensionalVector(b, b_flatten);
+  multiAddVectors(a_flatten, b_flatten, out);
+  return 0;
+}
+
+template <typename T>
+int Geometry::scaleVector(const T *a,
+                          const T factor,
+                          T *out,
+                          const int &dimension) {
+  for(int i = 0; i < dimension; ++i)
+    out[i] = a[i] * factor;
+  return 0;
+}
+
+template <typename T>
+int Geometry::scaleVector(const std::vector<T> &a,
+                          const T factor,
+                          std::vector<T> &out) {
+  out.resize(a.size());
+  return scaleVector(a.data(), factor, out.data(), a.size());
+}
+
+template <typename T>
+int Geometry::vectorProjection(const T *a,
+                               const T *b,
+                               T *out,
+                               const int &dimension) {
+  T dotProdBB = dotProduct(b, b, dimension);
+  if(dotProdBB > 1e-12) {
```

Comment:
instead of `1e-12`, you want to use one of the constants used in TTK (e.g. `1e-DBL_DIG` or `PREC_DBL`)

Reply:
you may want to double-check for other occurences of `1e-12`.

Reply:
Done in the next commit!

By the way, in the meantime I changed the function to adopt the convention used in the [Gram-Schmidt algorithm](https://en.wikipedia.org/wiki/Gram%E2%80%93Schmidt_process).
The convention is the following: the projection of a vector on a null vector is a null vector.
Therefore the function will gives a null vector in that case instead of returning failure (useful for the `gramSchmidt` function).